File: cpp/src/121.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <cassert>

using namespace std;


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_price = INT_MAX;
        int max_profit = 0;
        
        for (int price: prices) {
            if (price < min_price) 
                min_price = price;
            if (price - min_price > max_profit) 
                max_profit = price - min_price;
        }
        
        return max_profit;
    }
};


int main() {
    vector<int> in = {7, 1, 5, 3, 6, 4};
    Solution solution = Solution();
    int ret = solution.maxProfit(in);
    assert(ret == 5);

    in = {7, 6, 4, 3, 1};
    ret = solution.maxProfit(in);
    assert(ret == 0);

    cout << "Success" << endl;

    return 0;
}

File: go/src/121.go
func maxProfit(prices []int) int {
	n := len(prices)
	if n <= 1 {
		return 0
	}
	minPrice := prices[0]
	// dp[i] = max(dp[i-1], prices[i] - minPrice)
	dp := make([]int, n)
	dp[0] = 0
	for i := 1; i < n; i++ {
		dp[i] = max(dp[i-1], prices[i]-minPrice)
		minPrice = min(minPrice, prices[i])
	}

	return dp[n-1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}